# JavaScript Data Types<a id='toc0_'></a>

Hello friends, in this article, we will delve into data types and types in JavaScript. We will explore why data types and types are important and discuss the problems we may encounter if we don't use the correct data type or type for a variable. **In the following sections of the article, we will examine each data type and type separately.**

In this article, we'll cover:

- [Immutable and Mutable Concepts in JavaScript](#toc1_1_)
- [Data Type Concept in JavaScript](#toc1_2_)
  - [Data Types in JavaScript](#toc1_2_1_)
    - [Primitive Data Types](#toc1_2_1_1_)
    - [Object Data Types](#toc1_2_1_2_)
- [Summary](#toc1_3_)

I hope you enjoy reading.

Yazının Türkçe versiyonu için [linke](tr-js05-data-types.ipynb) tıklayabilirsiniz

<!-- vscode-jupyter-toc-config
	numbering=false
	anchor=true
	flat=false
	minLevel=1
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->


## <a id='toc1_1_'></a>[Immutable and Mutable Concepts in JavaScript](#toc0_)

**In JavaScript, variables are divided into two groups based on data types:**

- Immutable (Constant values)

- Mutable (Variable values)

Constant values are also referred to as **_Literals_**. Each assigned value has a new address in memory (RAM), and the content of the assigned value cannot be changed in memory.

**💡 We typically use constant values in places where we want to preserve the original data. For example, the original data may be used in multiple places. Unpreserved data can lead to undesired results within the program.**

**⚠️ Each time a new value is assigned to a constant variable, a new space is allocated in memory to store the data. This can lead to memory-related issues. Therefore, if performance is a priority, we should use constants less frequently.**

Variable values are also known as **_variables_**. The assigned value has an address in memory, and the content of the assigned value can be changed. **Therefore, they have the property of being references. In the case of a variable, a new space is not allocated in memory with each data assignment. The relevant reference address where the data is stored is found, and the old data is overwritten with the new data in the reference address.**

**💡 Variable values, with their use of reference addresses for data, do not occupy space in memory like constants. Therefore, if performance is a priority, we can use this data type.**

**Example**

Below is an illustration of the immutable property.


In [15]:
%%script node

// studentName variable is immutable.
let studentName = "Emin";

console.log(`Content of the studentName variable: ${studentName}`);

// Storing the content of studentName variable into the personName variable.
let personName = studentName;

/** 
 * Storing a new value in the studentName variable. In this case, a new address in RAM will be allocated for 
 * Hasan.
 */
studentName = "Hasan";

console.log(`Content of the studentName variable after assigning a new value: ${studentName}`);
console.log(`Pay attention here. Value stored in the personName variable: ${personName}`);


Content of the studentName variable: Emin
Content of the studentName variable after assigning a new value: Hasan
Pay attention here. Value stored in the personName variable: Emin


Let's create the same example using objects. Object properties have mutable characteristics.

**Example**


In [16]:
%%script node

// The studentName variable has mutable properties.
let studentName = { firstName: "Emin" };

console.log(`Content of the studentName variable: ${studentName["firstName"]}`);

// Storing the memory address of the studentName variable into the personName variable.
let personName = studentName;

/** 
 * Storing a new value in the studentName variable. In this case, a new address in RAM will not be allocated; the 
 * existing value will be overwritten with the new value.
 */
studentName["firstName"] = "Hasan";

console.log(`Content of the studentName variable after assigning a new value: ${studentName["firstName"]}`);
console.log(`Pay attention here. Value stored in the personName variable: ${personName["firstName"]}`);


Content of the studentName variable: Emin
Content of the studentName variable after assigning a new value: Hasan
Pay attention here. Value stored in the personName variable: Hasan


## <a id='toc1_2_'></a>[Data Type Concept in JavaScript](#toc0_)

Data types and data type concepts are crucial in JavaScript. To work with variables effectively, understanding data types and their concepts is essential.

A computer program cannot fully solve equations without knowing the data type or its concept, leading to incorrect results.

**Example**


In [1]:
%%script node

let x = 20 + "2020";

console.log(`Prints the expression to the console: ${x}`);

console.log(`Data type of the value 20: ${typeof 20}`);
console.log(`Data type of the value "2020": ${typeof "2020"}`);

Prints the expression to the console: 202020
Data type of the value 20: number
Data type of the value "2020": string


If we consider the above operation in mathematical terms, it is not logical. This is because the sum of the number 20 and the string "2020" would result in "202020". The reason for this is that the data types of the values are different; the number 20 is of the number data type, while the value "2020" is of the string data type.

**⚠️ To obtain correct results, variables must have the same data type.**

In the example above, let's convert the value "2020" to a number type and try again.

**Example**

In [3]:
%%script node

let x = 20 + Number(2020);

console.log(`Prints the expression to the console: ${x}`);

console.log(`Data type of the value 20: ${typeof 20}`);
console.log(`Data type of the value "2020": ${typeof 2020}`);

Prints the expression to the console: 2040
Data type of the value 20: number
Data type of the value "2020": number


As can be seen, the data type of both values is number. Therefore, the mathematical expression has yielded a logical result.

### <a id='toc1_2_1_'></a>[Data Types in JavaScript](#toc0_)

**JavaScript has two main data types:**

- Primitive (Primitive data types)

- Object (Reference data types)

Now let's classify them, my friends.


#### <a id='toc1_2_1_1_'></a>[Primitive Data Types](#toc0_)

In JavaScript, primitive data types are fundamental data structures that do not consist of complex structures and directly represent values.

Primitive data types in JavaScript are used in basic data operations and in representing values, forming the foundational building blocks of programs. They serve as the building blocks for complex objects and data structures.

**Features:**

- **Immutability:** Primitive data types are immutable once created. That is, once a primitive data type's value is assigned, it cannot be changed. This enhances data security and provides protection against unexpected side effects.

- **Value Type:** Primitive data types represent the values themselves, meaning values are stored directly in memory. When assigned to other variables, primitive data types are copied, so changes to one variable do not affect others.

- **Fixed Size in Memory:** Primitive data types have a fixed memory size. For example, a number is always stored in the same size of memory. This makes them more effective in terms of memory management.

- **Ease of Use:** Primitive data types are simple and can be used quickly. Operations and comparisons on primitive data types are performed quickly and efficiently.

- **Specific Value Ranges:** Each primitive data type represents a specific range of values. For instance, the **_number_** data type, being the most commonly used primitive data type in JavaScript, represents integers and floating-point numbers.

- **No Special Methods:** Primitive data types do not have special methods (functions) or properties, unlike objects and functions. Operations can be directly performed on primitive data types, and there are global functions (e.g., `parseInt`, `parseFloat`, `String`) that can be used for specific operations related to them.

**JavaScript has 6 primitive data types:**

1. String

2. Number

3. BigInt

4. Boolean

5. Undefined

6. Null

7. Symbol

**⚠️ The BigInt data type is a primitive data type in JavaScript due to its immutability in memory.**

**Example**

```javascript
// The following variable types have string characteristics.
let str1 = "Samet";
let str2 = "4";

// The following variable types have number characteristics.
let number1 = 10;
let number2 = 85.7;

// The following variable types have boolean characteristics.
let boolean1 = true;
let boolean2 = false;

// The following variable type has undefined characteristics.
let undefined1;
let x = undefined;

// The following variable type has null characteristics.
let null1 = null;

// The following variable type has symbol characteristics.
let symbol1 = Symbol("mySymbol");

// The following variable types have bigint characteristics.
let bigint1 = 1234567890123456789012345n;
let bigint2 = BigInt(1234567890123456789012345);
```


#### <a id='toc1_2_1_2_'></a>[Object Data Types](#toc0_)

In JavaScript, the object data type is a structure used to represent complex and flexible data structures. Objects are data structures containing key-value pairs and play a significant role in JavaScript programming. Objects represent a collection of properties and their assigned values enclosed in curly braces `{}`. **An object can also be created within square brackets `[]`, in which case, the object carries array-like properties.**

**Features:**

- **Mutable:** Object-typed variables with properties can be accessed in memory, and their contents can be updated. For example, when assigning a new value to an object-typed variable, a new address is not created in memory; instead, the new value overwrites the old one.

- **Key-Value Pairs:** JavaScript objects represent key-value pairs. These pairs define the properties and values of the object. Properties (keys) should be of string type, and values can be of any data type.

- **Dynamic and Extendable:** JavaScript objects are dynamic, meaning you can add, modify, or delete properties (keys) after the object is created. This makes objects flexible.

- **Properties and Methods:** JavaScript objects can contain both properties (key-value pairs) and methods (functions). Hence, objects can represent both data and functionality.

- **Reference Type:** Objects are reference-type data. When you assign an object to another variable, you're actually creating a copy of the reference, and both variables share the same object.

- **JSON (JavaScript Object Notation) Compatible:** JavaScript objects support data representation in JSON format. This is commonly used for sharing and exporting data.

- **Example-Based:** In JavaScript, objects are either class or prototype-based, and you can use class definitions or prototypes to create new objects.

**Object-typed data includes the following:**

- Object

- Function

- Array

- Date

**Example**

```javascript
function Test() {
  return "Function test";
}

// The following variable types have object characteristics.

// Object
let variable1 = { name: "Hasan", lastName: "Kaş" };

// Function
let variable2 = new Test();

// Array
let variable3 = ["Bursa", "Ankara", "İzmir"];

// Date
let variable4 = new Date("2023-10-26");
```


After discussing data types in JavaScript, let's now delve into the specifics of each data type.

In summary, each primitive data type in JavaScript has its own name for the data type.

**Example**


In [19]:
%%script node

// The following variables have a data type of string.
let str1 = "Samet";
let str2 = "4";

// The following variables have a data type of number.
let number1 = 10;
let number2 = 85.7;

// The following variables have a data type of boolean.
let boolean1 = true;
let boolean2 = false;

// The following variable has a data type of undefined.
let undefined1;
let x = undefined;

// The following variable has a data type of symbol.
let symbol1 = Symbol("mySymbol");

// The following variables have a data type of bigint.
let bigint1 = 1234567890123456789012345n;
let bigint2 = BigInt(1234567890123456789012345);

console.log(`The data type of str1 variable is: ${typeof str1}`);
console.log(`The data type of str2 variable is: ${typeof str2}`);

console.log(`The data type of number1 variable is: ${typeof number1}`);
console.log(`The data type of number2 variable is: ${typeof number2}`);

console.log(`The data type of boolean1 variable is: ${typeof boolean1}`);
console.log(`The data type of boolean2 variable is: ${typeof boolean2}`);

console.log(`The data type of undefined1 variable is: ${typeof undefined1}`);
console.log(`The data type of x variable is: ${typeof x}`);

console.log(`The data type of symbol1 variable is: ${typeof symbol1}`);

console.log(`The data type of bigint1 variable is: ${typeof bigint1}`);
console.log(`The data type of bigint2 variable is: ${typeof bigint2}`);


The data type of str1 variable is: string
The data type of str2 variable is: string
The data type of number1 variable is: number
The data type of number2 variable is: number
The data type of boolean1 variable is: boolean
The data type of boolean2 variable is: boolean
The data type of undefined1 variable is: undefined
The data type of x variable is: undefined
The data type of symbol1 variable is: symbol
The data type of bigint1 variable is: bigint
The data type of bigint2 variable is: bigint


**⚠️ This situation does not apply to the _null_ data type. This is because the data type of the _null_ value is of type object.**

**Example**


In [20]:
%%script node

let variable = null;

console.log(`The data type of the variable variable is: ${typeof variable}`);


The data type of the variable variable is: object


Objects are also created with the data type of object for data types.

**Example**


In [21]:
%%script node

function Test() { return "Function test"; }

// The data types of the following variables are of type object.

let variable1 = { name: "Hasan", lastName: "Kaş" };

let variable2 = new Test();

let variable3 = ["Bursa", "Ankara", "İzmir"];

let variable4 = new Date("2023-10-26");

console.log(`The data type of variable1 is: ${typeof variable1}`);
console.log(`The data type of variable2 is: ${typeof variable2}`);
console.log(`The data type of variable3 is: ${typeof variable3}`);
console.log(`The data type of variable4 is: ${typeof variable4}`);


The data type of variable1 is: object
The data type of variable2 is: object
The data type of variable3 is: object
The data type of variable4 is: object


## <a id='toc1_3_'></a>[Summary](#toc0_)

In this article, we focused on the data types and types in JavaScript. Variables are examined in two groups: **Immutable** and **Mutable**. Immutable values have a new address in memory, while mutable values point to the same address and their content can be changed.

There are six primitive types among data types: `String`, `Number`, `BigInt`, `Boolean`, `Undefined`, `Null`, and `Symbol`. Additionally, the `Object` type represents complex data structures.

In this article, we examined each data type with examples and emphasized the importance of working with data types correctly. In future articles, we will delve deeper into each type.
